### ISSUE TITLE
NGRAM_MATCH behavior does not appear to match documentation and makes it less useful for substring search

### ISSUE THREAD

**Issue by MrCreosote** [arangodb/arangodb#20118]
## My Environment

* __ArangoDB Version__:       3.11.4
* __Deployment Mode__:       Single server from `arangodb:3.11` Docker image from Dockerhub
* __Deployment Strategy__:   whatever the docker image does
* __Configuration__:                whatever the docker image does
* __Infrastructure__:               laptop
* __Operating System__:        Ubuntu 18.04
* __Total RAM in your machine__:        32GB, 20GB for VirtualBox VM
* __Disks in use__:        SSD
* __Used Package__:              Docker image

## Component, Query & Data

__Affected feature__:
ArangoSearch query using web interface

__AQL query (if applicable)__:
Query 1:
```
RETURN NGRAM_MATCH("flaviobacter", "obact", 1, "ngram3")
```

Query 2:
```
RETURN NGRAM_MATCH("xxobaxxxbacxxxactxxx", "obact", 1, "ngram3")
```
__AQL explain and/or profile (if applicable)__:
N/A

__Dataset__:
N/A

__Size of your Dataset on disk__:
N/A

__Replication Factor & Number of Shards (Cluster only)__:
RF: 3
Shards: N/A

## Steps to reproduce

1. Add the following analyzer to the DB:
```
{
    "name": "test::ngram3",
    "type": "pipeline",
    "features": [
        "frequency",
        "position"
    ],
    "properties": {
        "pipeline": [
            {
                "type": "norm",
                "properties": {
                    "locale": "en",
                    "case": "lower",
                    "accent": false
                }
            },
            {
                "type": "ngram",
                "properties": {
                    "min": 3,
                    "max": 3,
                    "preserveOriginal": false,
                    "streamType": "utf8",
                    "startMarker": "",
                    "endMarker": ""
                }
            }
        ]
    }
}
```
2. Execute query 1, note the `true` result
3. Execute query 2, note the `true` result

__Problem__:
The [documentation](https://docs.arangodb.com/3.11/aql/functions/arangosearch/#ngram_match) describes the scoring of the ngram match as:

> The similarity is calculated by **counting how long the longest sequence** of matching n-grams is, divided by the target’s total n-gram count. Only fully matching n-grams are counted. [Emphasis mine]

Based on that description I would expect query 1 to match (3 trigrams in the query, a sequence of 3 matching trigrams in the target for a score of 1) and query 2 to fail (3 trigrams in the query, the longest matching sequence is 1 trigram in the target for a score of 0.33), but both match.

I was hoping to use `NGRAM_MATCH` as a fast (and it's definitely really fast) substring query but given the current behavior that's not going to work, unless I'm doing something wrong.

__Expected result__:
Query 2's result should be `false` if I understand the documentation correctly

Thanks for any advice you can give me if there's something incorrect with my setup

**Comment by MBkkt** [comment]
Hello.

> query 2 to fail (3 trigrams in the query, the longest matching sequence is 1 trigram in the target for a score of 0.33)

It's not correct longest matching sequence is also 3 here
xx**oba**xxx**bac**xxx**act**xxx

**Comment by MBkkt** [comment]
In general you can imagine it like

"obact" -> target ngram array [oba bac act]
"xxobaxxxbacxxxactxxx" -> ngram array [xxo xob oba bax ....]

And we just find LCP longest common subsequence, when score is lcp divede by target array length

**Comment by MBkkt** [comment]
If you have long tokens (which is separated to ngrams) probably will be good is increase ngram size

**Comment by MrCreosote** [comment]
Thanks for clearing up my understanding.

> It's not correct, longest matching sequence is also 3 here
> xxobaxxxbacxxxactxxx

Hmm, that's not what I would think of as a sequence. I would think they'd need to be contiguous. I understand that things are working as intended though, so at most we're talking about a terminology difference.

That being said - is there a way to use arangosearch to find an arbitrary length substring in a field without scanning every document value for the field? There's `LIKE`, which the documentation says is backed by indexes, but for a search like `%foo%` to use an index you'd need a suffix tree or something like that, not just a regular inverted index (right?). My naive assumption is that inverted indexes will only accelerate `LIKE` searches where the left side of the query is anchored, like `foo%`.

**Comment by MBkkt** [comment]
> There's LIKE, which the documentation says is backed by indexes, but for a search like %foo% to use an index you'd need a suffix tree or something like that, not just a regular inverted index (right?). My naive assumption is that inverted indexes will only accelerate LIKE searches where the left side of the query is anchored, like foo%.

Yep, you're completely right, but we plan to add wildcard analyzer in 3.12. That helps for leading wildcard query too (internally it uses ngram and post filtering)

**Comment by MBkkt** [comment]
As another option you can try to make ngram with different size and search your substring by exact term search. But it can make inverted index big ofc

**Comment by MrCreosote** [comment]
> we plan to add wildcard analyzer in 3.12

Oh great, looking forward to 3.12 then. Hurry hurry hurry!
